### PR TITLE
Initial escalator implementation for folks to review

### LINKED ISSUE #155: Escalator
Implement a special handler type which will escalate any logs over a certain threshold to errors. This is useful for testing if your goal is to avoid introducing any extra warning noise to a production system (e.g., run the system under normal conditions and ensure that it doesn't produce any warnings).

### PR DESCRIPTION
**Pull request by rofinn**
Closes #155

### CODE REVIEW

**Comment by iamed2** [comment]
I like it, no notes from me for the first WIP commit

**Comment by rofinn** [comment]
A problem with this approach is that substituting the base logger won't do what we want for `@warn`

```
julia> using Memento

julia> Memento.config!("info"; substitute=true)
[notice | Memento]: Substituting global logging with Memento
Logger(root)

julia> push!(getlogger(), Escalator())
Escalator{DefaultFormatter}(DefaultFormatter("[{level} | {name}]: {msg}", Pair{Symbol,Bool}[Symbol("[") => 0, :level => 1, Symbol(" ") => 0, :| => 0, Symbol(" ") => 0, :name => 1, Symbol("]") => 0, :(:) => 0, Symbol(" ") => 0, :msg => 1], tz"America/Winnipeg"), Memento.Filter[Memento.Filter(Memento.var"#level_filter#27"{Escalator{DefaultFormatter}}(Escalator{DefaultFormatter}(#= circular reference @-4 =#)))], Dict{AbstractString,Int64}("trace" => 5,"critical" => 60,"error" => 50,"not_set" => 0,"notice" => 30,"alert" => 70,"info" => 20,"warn" => 40,"debug" => 10,"emergency" => 80…), "warn")

julia> @info "Hello World!"
[info | Main]: Hello World!

julia> @warn "Hello World!"
Exception handling log message: Memento.EscalationError("[warn | Main]: Hello World!")
  module=Main  file=REPL[8]  line=1
  Second exception: Memento.EscalationError("[error | Main]: Exception while generating log record in module Main at REPL[8]:1")
```

I think this is still better than nothing, but I'm not sure how to get around that without special casing the base logging behaviour.

**Comment by iamed2** [comment]
Maybe this needs to be a settable option on a logger instead? So that way it can be handled in `log`?

**Comment by rofinn** [comment]
Figured it out. Turns out I needed to add:

```
CoreLogging.catch_exceptions(logger::BaseLogger) = false
```

I suppose in the future we could do something similar for the Memento loggers, but for now that does what we want.

**Comment by rofinn** [comment]
```julia
julia> using Memento
[ Info: Precompiling Memento [f28f55f0-a522-5efc-85c2-fe41dfb9b2d9]

julia> Memento.config!("info"; substitute=true)
[notice | Memento]: Substituting global logging with Memento
Logger(root)

julia> push!(getlogger(), Escalator())
Escalator{DefaultFormatter}(DefaultFormatter("[{level} | {name}]: {msg}", Pair{Symbol,Bool}[Symbol("[") => 0, :level => 1, Symbol(" ") => 0, :| => 0, Symbol(" ") => 0, :name => 1, Symbol("]") => 0, :(:) => 0, Symbol(" ") => 0, :msg => 1], tz"America/Winnipeg"), Memento.Filter[Memento.Filter(Memento.var"#level_filter#27"{Escalator{DefaultFormatter}}(Escalator{DefaultFormatter}(#= circular reference @-4 =#)))], Dict{AbstractString,Int64}("trace" => 5,"critical" => 60,"error" => 50,"not_set" => 0,"notice" => 30,"alert" => 70,"info" => 20,"warn" => 40,"debug" => 10,"emergency" => 80…), "warn")

julia> @warn "Hello World!"
ERROR: Memento.EscalationError("[warn | Main]: Hello World!")
Stacktrace:
 [1] emit(::Escalator{DefaultFormatter}, ::DefaultRecord) at /Users/rory/repos/invenia/Memento.jl/src/handlers.jl:270
 [2] log(::Escalator{DefaultFormatter}, ::DefaultRecord) at /Users/rory/repos/invenia/Memento.jl/src/handlers.jl:44
 [3] log(::Logger, ::DefaultRecord) at /Users/rory/repos/invenia/Memento.jl/src/loggers.jl:362
 [4] #handle_message#70(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Base.CoreLogging.handle_message), ::Memento.BaseLogger, ::Base.CoreLogging.LogLevel, ::String, ::Module, ::String, ::Symbol, ::String, ::Int64) at /Users/rory/repos/invenia/Memento.jl/src/stdlib.jl:31
 [5] handle_message(::Memento.BaseLogger, ::Base.CoreLogging.LogLevel, ::String, ::Module, ::String, ::Symbol, ::String, ::Int64) at /Users/rory/repos/invenia/Memento.jl/src/stdlib.jl:29
 [6] top-level scope at logging.jl:320
```

File: src/handlers.jl [hunk 1]
```diff
@@ -219,3 +219,48 @@ function emit(handler::DefaultHandler{F, O}, rec::Record) where {F<:Formatter, O
...
+    levels::Dict{AbstractString, Int}
+    level::AbstractString
+end
+
+"""
+    Escalator(fmt=DefaultFormatter(); level="warn", levels=nothing)
+
+Escalates any logs it sees above a certain `level` to [`EscalationError`s](@ref).
+
+# Arguements
```
Comment (mattBrzezinski) [inline src/handlers.jl]:
```suggestion
# Arguments
```

File: src/handlers.jl [hunk 2]
```diff
@@ -219,3 +219,48 @@ function emit(handler::DefaultHandler{F, O}, rec::Record) where {F<:Formatter, O
...
+
+- `fmt::Formatter`: for converting `Record`s to error messages `Strings`
+
+# Keyword Arguements
```
Comment (mattBrzezinski) [inline src/handlers.jl]:
```suggestion
# Keyword Arguments
```

File: src/handlers.jl [hunk 3]
```diff
@@ -219,3 +219,48 @@ function emit(handler::DefaultHandler{F, O}, rec::Record) where {F<:Formatter, O
 
     flush(handler.io)
 end
+
+mutable struct Escalator{F} <: Handler{F}
+    fmt::F
+    filters::Array{Memento.Filter}
+    levels::Dict{AbstractString, Int}
+    level::AbstractString
```
Comment (mattBrzezinski) [inline src/handlers.jl]: I'm not familiar with `Memento`, but does it allow for creating a custom level? If not maybe we consider this making an enum so people can't pass in bad values to:

https://github.com/invenia/Memento.jl/pull/156/files#diff-46ed174b15e7a87714aaa0c48d1ef3a1R244

Comment (rofinn) [reply src/handlers.jl]: No, you can define custom levels in memento. It'll throw a key error if the level doesn't exist

Comment (iamed2) [reply src/handlers.jl]: It does allow custom levels

File: src/handlers.jl [hunk 4]
```diff
@@ -219,3 +219,48 @@ function emit(handler::DefaultHandler{F, O}, rec::Record) where {F<:Formatter, O
...
+getfilters(handler::Escalator) = handler.filters
+Base.push!(handler::Escalator, filter::Memento.Filter) = push!(handler.filters, filter)
+getlevels(handler::Escalator) = handler.levels
+getlevel(handler::Escalator) = handler.level
+function setlevel!(handler::Escalator, level::AbstractString)
+    handler.levels[level]     # Throw a key error if the levels isn't in levels
+    handler.level = level
+end
+
+emit(handler::Escalator, rec::Record) = throw(EscalationError(format(handler.fmt ,rec)))
```
Comment (iamed2) [inline src/handlers.jl]:
```suggestion
emit(handler::Escalator, rec::Record) = throw(EscalationError(format(handler.fmt, rec)))
```

**Pull request merged**